### DISCUSSION TITLE
How to Externalize Log4j2 Configuration for Multiple WARs in a Single Tomcat Instance

### DISCUSSION THREAD

**Discussion by maliksajidhussain** [apache/logging-log4j2#3028] [Q&A]
I understand that it's possible to externalize Log4j2 configuration files by specifying them outside the WAR. However, I have multiple WAR files deployed on a single Tomcat instance, each requiring its own separate Log4j2 configuration.

How can I configure Tomcat or pass Java arguments to ensure that each WAR uses its own external Log4j2 configuration file? What is the best approach to externalize the configuration for multiple WARs while maintaining separate logging configurations for each application?

Any insights or examples on how to handle this would be highly appreciated.

**Comment by ppkarwasz** [comment]
There are a couple of ways to have separate configuration files for each application:

1. You can place configuration files called `log4j2<contextName>.xml` on the classpath of **all** applications and configure a different `<contextName>` for each application. The default [Log4j Core initialization process](https://logging.apache.org/log4j/2.x/manual/configuration.html#automatic-configuration) will find those files.
2. You can use the [`log4j2.configurationFile`](https://logging.apache.org/log4j/2.x/manual/systemproperties.html#log4j2.configurationFile) Log4j configuration property together with [lookups](https://logging.apache.org/log4j/2.x/manual/lookups.html) so that the property evaluates to a different value for each web application.
3. You can use the [`log4jConfiguration`](https://logging.apache.org/log4j/2.x/jakarta.html#log4jConfiguration) Servlet initialization parameter recognized by the [`log4j-jakarta-web`](https://logging.apache.org/log4j/2.x/components.html#log4j-jakarta-web) web fragment.

### Setting the context name

The logger context name `<contextName>` is usually a hash of the web application classloader, so not really usable.
You can however change this in one of two ways:

1. You can add the [`log4j-jakarta-web` artifact](https://logging.apache.org/log4j/2.x/jakarta.html#log4j-jakarta-web) mentioned above to all your web applications. The artifact allows you to set `<contextName>` using the Servlet initialization parameter [`log4jContextName`](https://logging.apache.org/log4j/2.x/jakarta.html#log4jContextName). If you don't set it up explicitly, it has some reasonable defaults (also documented in the [`log4jContextName` description](https://logging.apache.org/log4j/2.x/jakarta.html#log4jContextName)).
2. You can use a **third-party** context selector that sets up the context name based on the Tomcat component hierarchy: [Tomcat Context Selector](https://oss.copernik.eu/tomcat/3.x/components/log4j-tomcat#TomcatContextSelector).

Of course my favorite approach is the one offered by the Copernik.eu extensions (I am the author).
There are a lot of subtleties, when dealing with logging in Tomcat (Tomcat JULI to Log4j API redirection, classloader delegation, etc.).
I created a [Logging in Tomcat using Log4j Core Guide](https://oss.copernik.eu/tomcat/3.x/guide) to deal with those altogether.

### Using lookups in `log4j2.configurationFile`

In the `log4j2.configurationFile` you can use lookups, so for example you can add:

```
-Dlog4j2.configurationFile=${catalina.base}/conf/log4j2-${web:contextPathName}.xml
```

could work.

**Reply by maliksajidhussain** [reply]
Hi @ppkarwasz Thanks for the reply. Let me try this.
